### PR TITLE
[SPARK-31521][CORE] Correct the fetch size when merging blocks into a merged block

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

Fix the wrong fetch size.

### Why are the changes needed?

The fetch size should be the sum of the size of merged block and the total size of those merging blocks. But we missed the size of merged block.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added a regression test.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan @xuanyuanking

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 1]
```diff
@@ -414,21 +414,23 @@ final class ShuffleBlockFetcherIterator(
...
-        }
+        val (startReduceId, size) =
+          if (mergedBlockInfo.nonEmpty && shouldMergeIntoPreviousBatchBlockId) {
+            // Remove the previous batch block id as we will add a new one to replace it.
+            val removed = mergedBlockInfo.remove(mergedBlockInfo.length - 1)
+              (removed.blockId.asInstanceOf[ShuffleBlockBatchId].startReduceId,
+                removed.size + toBeMerged.map(_.size).sum)
+          } else {
+            (startBlockId.reduceId, toBeMerged.map(_.size).sum)
+          }
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:425]: @Ngone51 . Could you make a unit test for this bug?

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 1 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:425]: ok, I'll try...

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:425]: Thank you, @Ngone51 !

**Comment by Ngone51** [comment]
retest this please.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 1 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:425]: Added

**Comment by Ngone51** [comment]
retest this please.

**Comment by Ngone51** [comment]
retest this please.

File: core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala [hunk 2]
```diff
@@ -1071,4 +1072,43 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
...
+      Int.MaxValue,
+      Int.MaxValue,
+      Int.MaxValue,
+      true,
+      false,
+      taskContext.taskMetrics.createTempShuffleReadMetrics(),
+      true)
+
+    val mergeMethod =
+      PrivateMethod[Seq[FetchBlockInfo]](Symbol("mergeContinuousShuffleBlockIdsIfNeeded"))
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala:1105]: shall we move this method to `object ShuffleBlockFetcherIterator` so it's easier to test?

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 3]
```diff
@@ -414,21 +414,23 @@ final class ShuffleBlockFetcherIterator(
...
-          mergedBlockInfo.remove(mergedBlockInfo.length - 1).blockId
-            .asInstanceOf[ShuffleBlockBatchId].startReduceId
-        } else {
-          startBlockId.reduceId
-        }
+        val (startReduceId, size) =
+          if (mergedBlockInfo.nonEmpty && shouldMergeIntoPreviousBatchBlockId) {
+            // Remove the previous batch block id as we will add a new one to replace it.
+            val removed = mergedBlockInfo.remove(mergedBlockInfo.length - 1)
+              (removed.blockId.asInstanceOf[ShuffleBlockBatchId].startReduceId,
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:421]: nit: indentation seems wrong

File: core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala (hunk 2 above)
Comment (Ngone51) [reply core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala:1105]: Sounds good.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @Ngone51 and @cloud-fan .
The new logic in functions and the outside numBlocksToFetch computation looks simpler than before. 
Merged to master/3.0.

**Comment by Ngone51** [comment]
thanks all!

**Pull request closed without merging**